include/oneapi/mkl/rng/device/detail/mcg31m1_impl.hpp

```diff
@@ -0,0 +1,238 @@
+// -1), but MCG31M1 needs only 2^31 - 1) if we want to do x % (2^N -1) we can
+// find out that: x = A + B * 2^N, where A = x % 2^N = x & 00..01..11 (binary)
+// where quantity of 1 is N, B = x / 2^N = x >> N also x = A + B * (2^N - 1 + 1)
+// = (A + B) + B * (2^N - 1), but (A + B) may be greater than (2^N - 1), that's
+// why we put x1 = A + B = A' + B' * 2^N = ... until new (A + B) < (2^N - 1) for
+// MCG31m1 N = 31
+template <typename T>
+static inline T custom_mod(std::uint64_t x) {
+    std::uint64_t b = x >> mcg31m1_param::bits;
+    std::uint64_t a = x & mcg31m1_param::m_64;
+    x = a + b;
+    b = x >> mcg31m1_param::bits;
+    a = x & mcg31m1_param::m_64;
+    return static_cast<T>(a + b);
+}
+
+template <std::int32_t VecSize>
+static inline sycl::vec<std::uint32_t, VecSize> custom_mod(
+    const sycl::vec<std::uint64_t, VecSize>& x) {
+    sycl::vec<std::uint64_t, VecSize> b = x >> mcg31m1_param::bits;
+    sycl::vec<std::uint64_t, VecSize> a = x & mcg31m1_param::m_64;
+    sycl::vec<std::uint64_t, VecSize> res = a + b;
+    b = res >> mcg31m1_param::bits;
+    a = res & mcg31m1_param::m_64;
+    res = a + b;
+    return res.template convert<std::uint32_t>();
+}
+
+static inline std::uint64_t power(std::uint64_t a, std::uint64_t n) {
+    std::uint64_t a2;
+    // initialize result by 1 for recurrence
+    std::uint32_t result = 1;
+
+    if (n == 0) {
+        // return (a^0)%m = 1
+        return std::uint64_t{ 1 };
+    }
+
+    // Recurrence loop
+    do {
+        // For each odd n
+        if (n & 1) {
+            a2 = static_cast<std::uint64_t>(result) * a;
+            result = custom_mod<std::uint32_t>(a2);
+        }
+        // n /= 2
+        n >>= 1;
+
+        a2 = a * a;
+        a = custom_mod<std::uint64_t>(a2);
+    } while (n);
+
+    return static_cast<std::uint64_t>(result);
+}
+
+template <std::int32_t VecSize>
+static inline void skip_ahead(engine_state<oneapi::mkl::rng::device::mcg31m1<VecSize>>& state,
+                              std::uint64_t num_to_skip) {
+    std::uint64_t loc_A = power(static_cast<std::uint64_t>(mcg31m1_param::a), num_to_skip);
+    state.s = custom_mod<std::uint32_t>(loc_A * static_cast<std::uint64_t>(state.s));
+}
+
+template <std::int32_t VecSize>
+static inline void init(engine_state<oneapi::mkl::rng::device::mcg31m1<VecSize>>& state,
+                        std::uint64_t n, const std::uint32_t* seed_ptr, std::uint64_t offset) {
+    if (n == 0)
+        state.s = 1;
+    else {
+        state.s = custom_mod<std::uint32_t>(seed_ptr[0]);
+        if (state.s == 0)
+            state.s = 1;
+    }
+    skip_ahead(state, offset);
+}
+
+template <std::int32_t VecSize>
+static inline sycl::vec<std::uint32_t, VecSize> generate(
+    engine_state<oneapi::mkl::rng::device::mcg31m1<VecSize>>& state) {
+    sycl::vec<std::uint64_t, VecSize> x(state.s);
+    sycl::vec<std::uint32_t, VecSize> res;
+#ifndef __HIPSYCL__
```

Comment:
Please add a small comment why we need a special branch for HIPSYCL

Reply:
Done